File: Merge_Sort/Objective-C/MergeSort.m
//
//  MergeSort.m
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

#import "MergeSort.h"

@implementation MergeSort

+ (NSArray<NSNumber *> *)sortedArray:(NSArray<NSNumber *> *)array order:(MergeSortOrder)order {
    NSUInteger arrayCount = array.count;

    if (arrayCount < 2) {
        return array.copy;
    }

    NSUInteger middleIndex = arrayCount / 2;

    NSArray *firstArray = [self sortedArray:[array subarrayWithRange:NSMakeRange(0, middleIndex)] order:order];
    NSArray *secondArray = [self sortedArray:[array subarrayWithRange:NSMakeRange(middleIndex, arrayCount - middleIndex)] order:order];

    return [self mergedFirstArray:firstArray secondArray:secondArray order:order];
}

+ (NSArray *)mergedFirstArray:(NSArray *)firstArray secondArray:(NSArray *)secondArray order:(MergeSortOrder)order {
    NSUInteger firstArrayCount = firstArray.count;
    NSUInteger secondArrayCount = secondArray.count;

    NSUInteger firstCounter = 0;
    NSUInteger secondCounter = 0;
    NSMutableArray *resultArray = [NSMutableArray arrayWithCapacity:firstArrayCount + secondArrayCount];

    while(firstCounter < firstArrayCount || secondCounter < secondArrayCount) {
        if (firstCounter == firstArrayCount) {
            [resultArray addObject:secondArray[secondCounter]];
            secondCounter++;
            continue;
        }

        if (secondCounter == secondArrayCount) {
            [resultArray addObject:firstArray[firstCounter]];
            firstCounter++;
            continue;
        }

        NSNumber *firstNumber = firstArray[firstCounter];
        NSNumber *secondNumber = secondArray[secondCounter];
        NSComparisonResult comparisonResult = [firstNumber compare:secondNumber];

        if (comparisonResult == NSOrderedSame) {
            [resultArray addObject:firstNumber];
            [resultArray addObject:secondNumber];

            firstCounter++;
            secondCounter++;
        }
        else if ((order == MergeSortOrderAsc && comparisonResult == NSOrderedAscending) ||
            (order == MergeSortOrderDesc && comparisonResult == NSOrderedDescending)) {
            [resultArray addObject:firstNumber];
            firstCounter++;
        }
        else {
            [resultArray addObject:secondNumber];
            secondCounter++;
        }
    }

    return resultArray.copy;
}

@end

File: Merge_Sort/Swift/MergeSort.swift
//
//  MergeSort.swift
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

import Foundation

struct MergeSort {

    enum SortOrder {
        case ASC
        case DESC
    }

    /** Complexity is O(n * log(n)) */
    static public func sort<T: Comparable>(_ array: [T?]?, order: SortOrder) -> [T?]? {
        let arrayCount = array?.count ?? 0

        if arrayCount < 2 {
            return array
        }

        let middleIndex = arrayCount / 2

        let firstArray = sort(Array(array![0..<middleIndex]), order: order)
        let secondArray = sort(Array(array![middleIndex..<arrayCount]), order: order)

        return merge(firstArray!, secondArray!, order: order);
    }

    static private func merge<T: Comparable>(_ firstArray: [T?], _ secondArray: [T?], order: SortOrder) -> [T?] {
        let firstArrayCount = firstArray.count
        let secondArrayCount = secondArray.count

        var firstCounter = 0
        var secondCounter = 0
        var resultArray: [T?] = []

        while firstCounter < firstArrayCount || secondCounter < secondArrayCount {
            if firstCounter == firstArrayCount {
                resultArray.append(secondArray[secondCounter])
                secondCounter += 1
                continue
            }

            if secondCounter == secondArrayCount {
                resultArray.append(firstArray[firstCounter])
                firstCounter += 1
                continue
            }

            let firstNumber = firstArray[firstCounter]
            let secondNumber = secondArray[secondCounter]

            if firstNumber == secondNumber {
                resultArray.append(firstNumber)
                resultArray.append(secondNumber)
                firstCounter += 1
                secondCounter += 1
            }
            else if (order == .ASC && (firstNumber == nil || (firstNumber != nil && secondNumber != nil && firstNumber! < secondNumber!))) ||
                (order == .DESC && (secondNumber == nil || (firstNumber != nil && secondNumber != nil && firstNumber! > secondNumber!))) {
                resultArray.append(firstNumber)
                firstCounter += 1
            }
            else {
                resultArray.append(secondNumber)
                secondCounter += 1
            }
        }

        return resultArray
    }

}
